lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -522,11 +583,55 @@ ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
     paramConstants.push_back(argValue);
   }
 
-  // TODO(constexpr patch): This is currently unused, so we don't need to
-  // calculate the correct value. Eventually, include code that calculates the
-  // correct value.
+  // If we reached an external function that hasn't been deserialized yet, make
+  // sure to pull it in so we can see its body.  If that fails, then we can't
+  // analyze the function.
+  if (callee->isExternalDeclaration()) {
+    callee->getModule().loadFunction(callee);
+    if (callee->isExternalDeclaration())
+      return computeOpaqueCallResult(apply, callee);
+  }
+
+  // Compute the substitution map for the callee, which maps from all of its
+  // generic requirements to concrete conformances and concrete types.
   SubstitutionMap calleeSubMap;
 
+  auto calleeFnType = callee->getLoweredFunctionType();
+  if (calleeFnType->getGenericSignature()) {
+    ApplySite AI(apply);
+
+    // Get the substitution map of the call.  This maps from the callee's space
+    // into the caller's world.
+    SubstitutionMap callSubMap;
+    if (calleeFnType->getRepresentation() ==
+        SILFunctionType::Representation::WitnessMethod) {
+      // Get the conformance out of the SILFunctionType and map it into our
+      // current type space.
+      auto conformance = calleeFnType->getWitnessMethodConformance();
+      auto selfTy = conformance.getRequirement()->getSelfInterfaceType();
+      auto conf = substitutionMap.lookupConformance(
+          selfTy->getCanonicalType(), conformance.getRequirement());
+      if (!conf.hasValue())
+        return evaluator.getUnknown((SILInstruction *)apply,
+                                    UnknownReason::Default);
+
+      // Witness methods have special magic that is required to resolve them.
+      callSubMap = getWitnessMethodSubstitutions(apply->getModule(), AI, callee,
+                                                 conf.getValue());
+    } else {
+      auto requirementSig = AI.getOrigCalleeType()->getGenericSignature();
+      callSubMap =
+          SubstitutionMap::get(requirementSig, apply->getSubstitutionMap());
```

Comment:
Just using apply->getSubstitutionMap() should be sufficient -- you don't have to build a new one since it should have the same generic signature.

Reply:
@slavapestov  Thanks for pointing this out. This copy can be removed.